### PR TITLE
Change the NodeResolver to return the symbol

### PR DESCRIPTION
**Pull request by pubudu91**
## Purpose
This PR changes the behaviour of the visitor used earlier to find the enclosing node for a given position to return the symbol itself.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java [hunk 1]
```diff
@@ -107,23 +103,14 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
...
-        } else if (node != null && (node.getKind() == USER_DEFINED_TYPE
-                || node.getKind() == NodeKind.UNION_TYPE_NODE
-                || node.getKind() == NodeKind.INTERSECTION_TYPE_NODE
-                || node.getKind() == NodeKind.VALUE_TYPE
-                || node.getKind() == NodeKind.BUILT_IN_REF_TYPE
-                || node.getKind() == NodeKind.CONSTRAINED_TYPE)) {
-            return Optional.ofNullable(
-                    SymbolFactory.createTypeDefinition(node.type.tsymbol, node.type.tsymbol.name.value));
+        SymbolFinder symbolFinder = new SymbolFinder();
+        BSymbol symbolAtCursor = symbolFinder.lookup(compilationUnit, position);
```
Comment (rasika) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:107]: Can we return an `Optional<BSymbol>` omitting the null check. Later in L113; we can use  a ternary for the return of `SymbolFactory.getBCompiledSymbol(...)` or `Optional.empty()`. As a further improvement we can return optional `Symbol` from the `SymbolFactory.getBCompiledSymbol(...)` as well.

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:107]: Not different from doing a null check right? And `SymbolFactory.getBCompiledSymbol()` should always be returning a `Symbol`, provided a valid `BSymbol`. Will update it to validate the `BSymbol`. Any filtering should be done before calling the factory method.

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/SymbolFactory.java [hunk 2]
```diff
@@ -92,7 +92,11 @@ public static Symbol getBCompiledSymbol(BSymbol symbol, String name) {
             return createTypeDefinition((BTypeSymbol) symbol, name);
         }
 
-        return null;
+        if (symbol == null) {
+            throw new IllegalArgumentException("Symbol is 'null'");
+        }
+
+        throw new IllegalArgumentException("Unsupported symbol kind: " + symbol.getClass().getName());
```
Comment (nadeeshaan) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/SymbolFactory.java:99]: Error message tells invalid kind, instead we show the class name. Shall we modify either the message or the input arg

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/SymbolFactory.java (hunk 2 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/SymbolFactory.java]: Changed kind -> type.

**Review by SupunS — APPROVED** [review]

**Review by nadeeshaan — APPROVED** [review]

**Pull request merged**